### ISSUE TITLE
How to inform the GC of the memory size of a given pointer?

### ISSUE THREAD

**Issue by ioquatix** [ffi/ffi#753]
In some cases, e.g. https://github.com/libvips/ruby-vips/issues/67#issuecomment-603048609, it's entirely possible that a pointer to an opaque data structure represents a huge amount of data, e.g. a 30MiB pixel buffer, etc.

But this cannot be communicated to the Ruby GC, and so the GC won't work correctly in order to free such allocations quickly.

My understanding is we need to use something like `rb_typed_data_t` and specify a memsize callback.

**Comment by larskanis** [comment]
Makes sense, although the memsize callback isn't used by Ruby's GC, but only by `ObjectSpace.memsize_of(obj)` . The GC must be notified explicit by calling `rb_gc_adjust_memory_usage()` if the memory is allocated outside of FFI. I implemented something very similar in ruby-pg: https://github.com/ged/ruby-pg/commit/c0cad36979a27bf25ef832e51c58dc530ae8b734

Ruby should already count memory that is allocated by FFI::MemoryPointer since it's using ruby's `xmalloc()`. But FFI doesn't know about memory that is not allocated by FFI, so that it can not count it. Maybe we could provide a function which allows the FFI user to register memory on it's own and pass it through to `rb_gc_adjust_memory_usage()`.

**Comment by ioquatix** [comment]
It makes total sense to expose this given how critical it is to GC behaving correctly, and how common it happens in FFI situations.

I did a nasty hack/proof of concept:

```
module GC
	extend FFI::Library
	ffi_lib FFI::CURRENT_PROCESS
	
	attach_function :rb_gc_adjust_memory_usage, [:ssize_t], :void
end

def acquire(img)
	GC.rb_gc_adjust_memory_usage(36*1024*1024)
	ObjectSpace.define_finalizer(img, self.method(:release))
end

def release(amount)
	GC.rb_gc_adjust_memory_usage(36*1024*1024)
end

img = Vips::Image.new_from_file filename, :access => :sequential
acquire(img)
		
img.write_to_file 'test.jpg', :Q => 50
```

We can probably do something better in FFI :)

**Comment by JasonLunn** [comment]
Has this issue been solved another way? If so, has anyone solved this problem generally for realloc'ing as well as malloc'ing?
If not, has anyone seen a JRuby equivalent of the proof-of-concept above?

CC: @headius
